src/lib.rs

```diff
@@ -215,6 +215,37 @@ pub fn narrowable_rboehm(args: TokenStream, input: TokenStream) -> TokenStream {
                 unsafe { gc.assume_init() }
             }
 
+
+            /// Create a narrow pointer to #trait_id. `layout` must be at least big enough for an
+            /// object of type `U` (but may optionally be bigger) and must have at least the same
+            /// alignment that `U requires (but may optionally have a bigger alignment). `init`
+            /// will be called with a pointer to uninitialised memory into which a fully
+            /// initialised object of type `U` *must* be written. After `init` completes, the
+            /// object will be considered fully initialised: failure to fully initialise it causes
+            /// undefined behaviour. Note that if additional memory was requested beyond that
+            /// needed to store `U` then that extra memory does not have to be initialised after
+            /// `init` completes.
+            pub unsafe fn new_from_layout<U: #trait_id, F>(layout: ::std::alloc::Layout,
+                init: F) -> ::rboehm::Gc<Self>
+                where F: FnOnce(*mut U)
+            {
+                let (layout, uoff) = ::std::alloc::Layout::new::<usize>().extend(layout).unwrap();
+                // Check that we've not been given an object whose alignment
+                // exceeds that of a usize.
+                debug_assert_eq!(uoff, ::std::mem::size_of::<usize>());
```

Comment:
Is there a reason this isn't allowed? The rust doc seems to suggest that the layout can be optionally bigger than the alignment requirements of `U`.

Reply:
This is an old thing that's just been moved over in this PR to ensure that we store the vtable OK. I *think* it can now be relaxed, but it's probably worth doing a thorough analysis of this in another PR.

Reply:
Sure. For now shall we document this restriction in the rust doc since it seems to be contradictory at the moment?

Reply:
Let me have another look to remind myself why this is there.

Reply:
Hmm, this is going to require more thought than I first expected: I think this is my attempt to avoid having to dynamically read alignment. I suggest that we keep the comment as-is but I raise an issue to fix this if that's OK?

Reply:
That works for me